src/DynamoCoreWpf/Views/Search/LibraryView.xaml.cs

```diff
@@ -180,19 +184,48 @@ private void ExpandCategory(TreeViewItem sender, NodeCategoryViewModel selectedC
             foreach (var categoryToBeCollapsed in allExpandedCategories)
             {
                 // If class button was clicked.
-                if (categoryToBeExpanded == null)
+                if (selectedClass.SubCategories.Count == 0)
                 {
                     var categoryClasses = categoryToBeCollapsed.Items[0] as ClassesNodeCategoryViewModel;
                     // Ensure, that this class is not part of current category.
                     if (categoryClasses != null)
-                        categoryToBeCollapsed.IsExpanded = categoryClasses.Items.Contains(selectedClass);
+                    {
+                        if (categoryClasses.Items.Contains(selectedClass))
+                        {
+                            categoryToBeCollapsed.IsExpanded = true;
+                            // Selected class has been found! Collapse all other sub categories.
+                            foreach (var ele in categoryToBeCollapsed.SubCategories.Where(cat => cat.IsExpanded == true))
+                                ele.IsExpanded = false;
+                        }
+                        else
+                            // If there is no sub categories, we can be sure, 
+                            // that selected class is not inside.
+                            if (categoryToBeCollapsed.SubCategories.Count == 0)
+                                categoryToBeCollapsed.IsExpanded = false;
+                            // If there are some sub categories, we have to check:
+                            // is selected class inside of one of these sub categories?
+                            else
+                                categoryToBeCollapsed.IsExpanded = ExpandCategory(categoryToBeCollapsed.Items.OfType<NodeCategoryViewModel>(), selectedClass);
+                    }
+                    // If categoryClasses == null, that means there are no classes. 
+                    // It can be when namespace is inside of another namespace.
+                    else
+                    {
+                        categoryToBeCollapsed.IsExpanded = ExpandCategory(categoryToBeCollapsed.Items.OfType<NodeCategoryViewModel>(), selectedClass);
+                    }
                 }
                 // If category button was clicked.
                 else
                 {
-                    categoryToBeCollapsed.IsExpanded = false;
```

Comment:
The main problem was here.
categoryToBeCollapsed.IsExpanded = false; works fine only, if class and namespace are at the same level.
E.g.
-Root category
---Class
---Namespace1
---Namespace2
Then we go through all expanded namespaces and close them.
But it doesn't work, when class and namespace are the different levels.
E.g.
-Root category
---Class
---Namespace1
-------Namespace2
I created recursion to get this behavior.